### PR TITLE
Test fiat #404

### PR DESCRIPTION
**Pull request by ReubenHill**

### CODE REVIEW

**Review by dham — APPROVED** [review]

**Comment by ReubenHill** [comment]
I test the following
```python
from firedrake import *
m = ExtrudedMesh(UnitSquareMesh(2, 3), 2)
V = FunctionSpace(m, "CG", 2)
f = Function(V)
f.at([0.5, 0.5, 0.4])
```
With fiat master, our kernel has
```c
static inline void to_reference_coords_kernel(void *result_, double *x0, int *return_value, double *C)
{
    ...
    // Are we inside the reference element?
    *return_value = PetscRealPart(X[0]) + 1.0e-14 >= 0 && PetscRealPart(X[1]) + 1.0e-14 >= 0 && PetscRealPart(X[2]) + 1.0e-14 >= 0 && PetscRealPart(X[2]) - 1.0e-14 <= 1 && PetscRealPart(X[0]) + PetscRealPart(X[1]) - 1.0e-14 <= 1;
}
```
(16 operations, all adds and logical)

With branch `[ReubenHill/point-locate](https://github.com/firedrakeproject/fiat/tree/ReubenHill/point-locate)` (https://github.com/firedrakeproject/fiat/pull/33) we now have
```
```c
static inline void to_reference_coords_kernel(void *result_, double *x0, int *return_value, double *C)
{
    ...
    // Are we inside the reference element?
    *return_value = 0.5*fabs(PetscRealPart(X[2])) + 0.5*fabs(PetscRealPart(X[2]) - 1.0) - 0.5 <= 1.0e-14 && 0.5*fabs(PetscRealPart(X[0])) + 0.5*fabs(PetscRealPart(X[1])) + 0.5*fabs(PetscRealPart(X[0]) + PetscRealPart(X[1]) - 1.0) - 0.5 <= 1.0e-14;
}
```
(5 multiplications, 5 fabs, 8 add/sub, 3 logical)

**Comment by ReubenHill** [comment]
If we apply the following to  `[ReubenHill/point-locate](https://github.com/firedrakeproject/fiat/tree/ReubenHill/point-locate)`/(https://github.com/firedrakeproject/fiat/pull/33)
```diff
diff --git a/FIAT/reference_element.py b/FIAT/reference_element.py
index fa4d47b..c4e880b 100644
--- a/FIAT/reference_element.py
+++ b/FIAT/reference_element.py
@@ -657,7 +657,7 @@ class UFCSimplex(Simplex):
         # sum(bary-abs(bary)) = -4.
         # - 0.5 * sum(bary-abs(bary)) = 2.0
         # which is the correct L1 distance from the cell to the point.
-        return - 0.5 * sum([b - abs(b) for b in bary])
+        return abs(0.5 * sum([b - abs(b) for b in bary]))
 
 
 class Point(Simplex):
```
(i.e. swap `-` for `abs`) we get

```c
static inline void to_reference_coords_kernel(void *result_, double *x0, int *return_value, double *C)
{
    ...
    // Are we inside the reference element?
    *return_value = fabs(0.5*fabs(PetscRealPart(X[2])) + 0.5*fabs(PetscRealPart(X[2]) - 1.0) - 0.5) <= 1.0e-14 && fabs(0.5*fabs(PetscRealPart(X[0])) + 0.5*fabs(PetscRealPart(X[1])) + 0.5*fabs(PetscRealPart(X[0]) + PetscRealPart(X[1]) - 1.0) - 0.5) <= 1.0e-14;
}
```
(5 multiplications, 7 fabs, 8 add/sub, 3 logicals) which is much of a muchness

File: firedrake/pointquery_utils.py [hunk 1]
```diff
@@ -65,8 +66,7 @@ def is_affine(ufl_element):
 def inside_check(fiat_cell, eps, X="X"):
     dim = fiat_cell.get_spatial_dimension()
     point = tuple(sympy.Symbol("PetscRealPart(%s[%d])" % (X, i)) for i in range(dim))
-
-    return " && ".join("(%s)" % arg for arg in fiat_cell.contains_point(point, epsilon=eps).args)
+    return ccode(fiat_cell.contains_point(point, epsilon=eps))
```
[the remark this answers is not in the document]
Comment (ReubenHill) [reply firedrake/pointquery_utils.py:69]: I'll stick it in its own PR

[the remark this answers is not in the document]
Comment (ReubenHill) [reply firedrake/pointquery_utils.py:69]: https://github.com/firedrakeproject/firedrake/pull/2741

**Pull request closed without merging**